214 Vehicle Detection And Recognition with OpenVINO

IRONICBo opened the pull request:
This is a notebook using detection model and classification model to detect vehicle and recognize the attributes of each vehicles.

The flowcharts is below:

![vehicle-inference-flow](https://user-images.githubusercontent.com/47499836/157867076-9e997781-f9ef-45f6-9a51-b515bbf41048.png)

And the final result is here.
Resource:
![resource](https://user-images.githubusercontent.com/47499836/163544843-3e4f12cb-cf92-4633-829e-372dfe45bdd8.png)
Result:
![output](https://user-images.githubusercontent.com/47499836/163544861-fa2ad64b-77df-4c16-b065-79183e8ed964.png)

adrianboguszewski commented:
Thanks, @IRONICBo. Is this in progress or finished and ready for review?

IRONICBo commented:
> Thanks, @IRONICBo. Is this in progress or finished and ready for review?

Thank you for your attention,the notebook is finished now and need CR~

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/README.md:
```diff
@@ -0,0 +1,19 @@
+
+## Notebook Contents
+
+In this notebook, we will use both detection model and classification model with OpenVINO.We use [Object Detection Models](https://docs.openvino.ai/2020.2/usergroup3.html) and [Object Recognition Models](https://docs.openvino.ai/2020.2/usergroup4.html) from [Open Model Zoo](https://github.com/openvinotoolkit/open_model_zoo).Detection model is used to detect vehicle position.Besides, we crop singl ...
```
The links redirect to the old documentation page. It's better to replace them by open model zoo links e.g. https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/vehicle-detection-0200

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/README.md:
```diff
@@ -0,0 +1,19 @@
+# Vehicle Detection And Recognition with OpenVINO
+
+[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/IRONICBo/openvino_notebooks/HEAD?labpath=notebooks%2F214-vehicle-detection-and-recognition%2F214-vehicle-detection-and-recognition.ipynb)
```
As the notebook is not merged yet, you can remove the Binder link

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/README.md:
```diff
@@ -0,0 +1,19 @@
+## Notebook Contents
+
+In this notebook, we will use both detection model and classification model with OpenVINO.We use [Object Detection Models](https://docs.openvino.ai/2020.2/usergroup3.html) and [Object Recognition Models](https://docs.openvino.ai/2020.2/usergroup4.html) from [Open Model Zoo](https://github.com/openvinotoolkit/open_model_zoo).Detection model is used to detect vehicle position.Besides, we crop singl ...
+<div  align='center' ><img src="data/vehicle-inference-flow.png" alt="drawing" width="1000"/></div>
```
Instead of html here, please use markdown

adrianboguszewski commented on notebooks/README.md:
```diff
@@ -97,6 +100,7 @@ More amazing notebooks here! Please take a look of the complete list.
 <p>
 <details>
 <summary> 🏃 Model Training</summary>
+>>>>>>> 605687251832dc30704174785d2e9f1642cdd65f
```
Residues after merging

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "\n",
+    "Finally, we will get the result:\n",
+    "\n",
+    "<div  align='center' ><img src=\"https://user-images.githubusercontent.com/47499836/157867020-99738b30-62ca-44e2-8d9e-caf13fb724ed.png\" alt=\"drawing\"/></div>\n"
```
The same like in README

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,545 @@
+   "source": [
+    "# Vehicle Detection And Recognition with OpenVINO\n",
+    "\n",
+    "In this notebook, we will use both detection model and classification model with OpenVINO.We use [Object Detection Models](https://docs.openvino.ai/2020.2/usergroup3.html) and [Object Recognition Models](https://docs.openvino.ai/2020.2/usergroup4.html) from [Open Model Zoo](https://github.com/openvinotoolkit/open_model_zoo).Detection model is used to detect vehicle position.Besides, we crop  ...
```
The same like in README

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "\n",
+    "We need to download pretrained models to continue our progress.We use `omz_downloader`, a command-line tool installed by `openvino-dev` package.\n",
+    "\n",
+    "> Note: If you want to change the model,you need to modify the model name.If you want to change the precision,you need to modify the precision value. "
```
If you know what else models one can use here, please provide the names

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,545 @@
+   "outputs": [],
+   "source": [
+    "# Initialize inference engine\n",
+    "ie_core = IECore()\n",
```
You're using the old OpenVINO API. It was changed just after you opened this PR. If you can please update to the new API.

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "    plt.imshow(raw_image)\n",
+    "\n",
+    "\n",
+    "def softmax(x: np.ndarray) -> np.ndarray:\n",
```
You don't need this function. As stated here: https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/vehicle-attributes-recognition-barrier-0039 the model has its own softmax inside.

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "            # In case that bounding box is found at the top of the image, \n",
+    "            # we position upper box bar little bit lower to make it visible on image \n",
+    "            (x_min, y_min, x_max, y_max) = [\n",
+    "                int(max(corner_position * ratio_y * 256, 10)) if idx % 2 \n",
```
Please use input size variables instead of magic numbers (256)

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "# Three vehicles in car.jpg\n",
+    "# test_car = resized_image_de[100:164, 0:100]\n",
+    "# test_car = resized_image_de[130:215, 166:254]\n",
+    "test_car = resized_image_de[92:152, 126:209]\n",
```
Please use variables like boxes[0] instead of magic numbers

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,545 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def convert_result_to_image(exec_net_re, bgr_image, resized_image, boxes, threshold=0.6):\n",
```
This function is almost the same as the previous one. Please avoid code repetition

adrianboguszewski commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "            # parameters in putText function are: img, text, org, fontFace, fontScale, color, thickness, lineType\n",
+    "            rgb_image = cv2.putText(\n",
+    "                rgb_image, \n",
+    "                f\"A {attr_color} {attr_type}\",\n",
```
You can remove this "A"

adrianboguszewski reviewed (changes requested):
Thanks, @IRONICBo! We really appreciate your work. Unfortunately, we switched to the new API, just a few hours after you opened this PR, so there is a need to update this notebook too. It should be rather straightforward. I left some comments in the code. And please add a link to this notebook to the main README file as well.

IRONICBo commented:
> Thanks, @IRONICBo! We really appreciate your work. Unfortunately, we switched to the new API, just a few hours after you opened this PR, so there is a need to update this notebook too. It should be rather straightforward. I left some comments in the code. And please add a link to this notebook to the main README file as well.

Thank you for your suggestions!!! I have changed to new API and fixed the irregularities in my pull request.Also, I added the  notebook link to main README and README_cn.

Debskij commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,518 @@
+   "source": [
+    "# Load Models\n",
+    "\n",
+    "In this notebook,we will need detection model and recognition model.When we download models,we need to initialize inference engine(IECore),and use `read_network` to read network architecture and weights from *.xml and *.bin files.Then,we load the network on the \"CPU\" using `load_network`."
```
Please adjust comments to new API

Debskij replied:
```suggestion
    "In this notebook, we will need detection model and recognition model. When we download models, we need to initialize inference engine(IECore),and use `read_network` to read network architecture and weights from *.xml and *.bin files. Then, we load the network on the \"CPU\" using `load_network`."
```

Please add spaces after `.` and `,` in comments

Debskij commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,518 @@
+    "    # compile the model for the CPU (you can use GPU or MYRIAD as well)\n",
+    "    compiled_model = ie_core.compile_model(model=model, device_name=\"CPU\")\n",
+    "    # Get input and output names of nodes\n",
+    "    input_keys = compiled_model.inputs[0]\n",
+    "    output_keys = compiled_model.outputs[0]\n",
```
```suggestion
    "    input_keys = compiled_model.input(0)\n",
    "    output_keys = compiled_model.output(0)\n",
```

Debskij commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,518 @@
+   "source": [
+    "### Get attributes from model\n",
+    "\n",
+    "We use `XXX.input_info.tensor_desc.dims` to get data shapes"
```
Adjust comment to new api

Debskij commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,518 @@
+   "outputs": [],
+   "source": [
+    "# Inference the network\n",
+    "request_de = compiled_model_de.create_infer_request()\n",
+    "request_de.infer(inputs={input_key_de.any_name: input_image_de})\n",
+    "\n",
+    "# Get the result\n",
+    "boxes = request_de.get_tensor(output_keys_de).data\n",
```
```suggestion
    "boxes = compiled_model_de([input_image_de])[output_keys_de]\n",
```

Debskij commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,518 @@
+    "    input_image_re = np.expand_dims(resized_image_re.transpose(2, 0, 1), 0)\n",
+    "    \n",
+    "    # inference the recognition net\n",
+    "    request_re = compiled_model_re.create_infer_request()\n",
+    "    request_re.infer(inputs={input_key_re.any_name: input_image_re})\n",
+    "    \n",
+    "    # predict result\n",
+    "    predict_colors = request_re.get_tensor(\"color\").data\n",
+    "    # delete the dim of 2, 3\n",
+    "    predict_colors = np.squeeze(predict_colors, (2, 3))\n",
+    "    predict_types = request_re.get_tensor(\"type\").data\n",
+    "    predict_types = np.squeeze(predict_types, (2, 3))\n",
+    "\n",
```
Please use similar api simplification as suggested above

Debskij commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,518 @@
+   "source": [
+    "### Conbine two models\n",
+    "\n",
+    "Congratulations!Now we can use detection model to crop single vehicle and recognize the vehicle's attribute."
```
```suggestion
    "Congratulations! Now we can use detection model to crop single vehicle and recognize the vehicle's attribute."
```

Debskij commented on notebooks/214-vehicle-detection-and-recognition/214-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,518 @@
+   "source": [
+    "### Recognize the vehicle's attributes\n",
+    "\n",
+    "We choose one of the box,and crop the vehicle to test recognition model.Similarly,we need to resize the input image, and infer it."
```
```suggestion
    "We choose one of the box, and crop the vehicle to test recognition model. Similarly, we need to resize the input image, and infer it."
```

adrianboguszewski commented:
Thanks @IRONICBo. It's almost ready. Please change the number of the notebook to 218 as we have already merged 214. Please change also the link in README, notebooks/README and fix spaces in the first paragraph.

IRONICBo commented:
> Thanks @IRONICBo. It's almost ready. Please change the number of the notebook to 218 as we have already merged 214. Please change also the link in README, notebooks/README and fix spaces in the first paragraph.

Thanks! @adrianboguszewski I have changed the notebook from 214 to 218 and fixed the spaces in my pull request. :)

paularamo commented on notebooks/218-vehicle-detection-and-recognition/218-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,514 @@
+   "source": [
+    "# Vehicle Detection And Recognition with OpenVINO\n",
+    "\n",
+    "In this notebook, we will use both detection model and classification model with OpenVINO. We use [Object Detection Models](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/vehicle-detection-0200) and [Object Recognition Models](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/vehicle-attributes-recognition-barrier-0039) from [Intel Open Mo ...
```
```suggestion
    "In this notebook, we will use both detection model and classification model with OpenVINO. We use [Object Detection Models](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/vehicle-detection-0200) and [Object Recognition Models](https://github.com/openvinotoolkit/open_model_zoo/tree/master/models/intel/vehicle-attributes-recognition-barrier-0039) from [Intel Open Model Zoo](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/intel/index.md). Detection model is used to detect vehicle position. Besides, we crop single vehicle and infer with classification model to recognize attributes of single vehicle. The pipline is here:\n"
    "\n",
    "\n",    
```

paularamo reviewed (approved):
@IRONICBo Thanks for your contribution, please check the conflicts in your branch.

@ryanloney This notebook is ready for your review.

paularamo commented on README_cn.md:
```diff
@@ -55,6 +55,7 @@
 | [211-speech-to-text](notebooks/211-speech-to-text/)<br>[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/openvinotoolkit/openvino_notebooks/HEAD?filepath=notebooks%2F211-speech-to-text%2F211-speech-to-text.ipynb) | 声音到文字的识别模型推理 | <img src="https://user-images.githubusercontent.com/36741649/140987347-279de058-55d7-4772-b013-0f2b12deaa61.png" width=225>|
 | [212-onnx-style-transfer](notebooks/212-onnx-style-transfer/)<br>[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/openvinotoolkit/openvino_notebooks/HEAD?filepath=notebooks%2F212-onnx-style-transfer%2F212-onnx-style-transfer.ipynb) | 通过神经风格转换将图像转换为五种不同的风格 | <img src="https://user-images.githubusercontent.com/77325899/147358090-ff5b21f5-0efb-4aff-8444-9d07add49b92.png" ...
 | [213-question-answering](notebooks/213-question-answering/)<br>[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/openvinotoolkit/openvino_notebooks/HEAD?filepath=notebooks%2F213-question-answering%2F213-question-answering.ipynb) | 根据语境回答问题 | <img src="https://user-images.githubusercontent.com/4547501/152571639-ace628b2-e3d2-433e-8c28-9a5546d76a86.gif" width=225> |
+| [218-vehicle-detection-and-recognition](notebooks/218-vehicle-detection-and-recognition/)<br> | 车辆检测与属性识别 | <img src = "https://user-images.githubusercontent.com/47499836/157867020-99738b30-62ca-44e2-8d9e-caf13fb724ed.png" width=225> |
```
```suggestion
| [218-vehicle-detection-and-recognition](notebooks/218-vehicle-detection-and-recognition/)<br> | 车辆检测与识别 | <img src = "https://user-images.githubusercontent.com/47499836/157867020-99738b30-62ca-44e2-8d9e-caf13fb724ed.png" width=225> |
```

paularamo replied:
@IRONICBo In order to centralize data, please use data (images/video) that you could find here. You don't need to download that data, just use the URL instead of the data folder.

https://storage.openvinotoolkit.org/data/test_data/images/person-bicycle-car-detection.bmp
https://storage.openvinotoolkit.org/data/test_data/videos/car-detection.mp4
https://storage.openvinotoolkit.org/data/test_data/videos/person-bicycle-car-detection.mp4 

Use:
`!curl -o person-vehicle-bike.mp4 https://storage.openvinotoolkit.org/data/test_data/videos/person-bicycle-car-detection.mp4
`
@adrianboguszewski this is applicable for other PRs.

IRONICBo replied:
@paularamo Thank you for your suggestion. I use the githubusercontent URL  instead of  the data folder.

paularamo replied:
@IRONICBo Thanks. Yes, Please change the image to one of the references, and avoid putting the data folder, in order to use less space per notebook and avoid intellectual property issues on new images we include.

IRONICBo replied:
@paularamo Ok, I have changed the image to https://storage.openvinotoolkit.org/data/test_data/images/person-bicycle-car-detection.bmp and remove the data folder.

Besides, I'll try to do model inference on video like https://storage.openvinotoolkit.org/data/test_data/videos/car-detection.mp4.

IRONICBo replied:
@paularamo Thank you for your advice. I read the notebook of https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/utils/notebook_utils.ipynb#L208 and I changed the "curl" with "download_file" function, And I it works well.

zhuo-yoyowz commented on notebooks/218-vehicle-detection-and-recognition/218-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,513 @@
+   "source": [
+    "##### Recognition progressing\n",
+    "\n",
+    "We will get the result contains colors(White, gray, yellow, red, green, blue, black)and types(Car, bus, truck, van), then we need to get the probability of each attribute, finally we need to choose the max probability as the result."
```
```suggestion
    "We will get the result contains colors (White, gray, yellow, red, green, blue, black) and types (Car, bus, truck, van), then we need to get the probability of each attribute, finally we need to choose the max probability as the result."
```

zhuo-yoyowz commented on notebooks/218-vehicle-detection-and-recognition/218-vehicle-detection-and-recognition.ipynb:
```diff
@@ -0,0 +1,513 @@
+   "source": [
+    "### Recognize the vehicle's attributes\n",
+    "\n",
+    "We choose one of the box, and crop the vehicle to test recognition model. Similarly, we need to resize the input image, and infer it."
```
```suggestion
    "We choose one of the detected boxes, and crop the vehicle to test recognition model. Similarly, we need to resize the input image, and infer it."
```

IRONICBo commented:
@zhuo-yoyowz Thank your for your review, I have updated my notebook. :）

ryanloney commented:
@Debskij @paularamo @adrianboguszewski let's discuss on Wednesday. Close to being ready to merge, but need to align with OV 2.0 API naming for runtime and not use IE references.
